### PR TITLE
Move from gcr to ghcr for the router and examples

### PR DESCRIPTION
**Pull request by dolfolife**
## Summary

The intent of these changes is to move the registry from GCR to GHCR. 

### TODO
- [X] Router
- [x] Examples
- [x] Update the `veba.sbom` file with the new urls

## Change Type

What type of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the types of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [X] Build related changes
- [ ] Documentation changes
- [X] Other (please describe): Image location

## Additional Context

I do not have much context where most of the images are built. I took a look at the Github Action and found that the router is built there but none of the examples are. I am working on deleting the dependency of VEBA from GCR and moving to the GitHub registry since it is free for public repositories.

### CODE REVIEW

**Comment by dolfolife** [comment]
@lamw I am ready to review these changes. 

The main review would be to push all examples and the router using ghcr to this repository using the documentation I added in the PR.

Notes:
- There was a `FROM lamw/<repo>` in one of the dockerfile on the example templates of a private repo that I deleted to make the build work.
- Some of the python examples had a conflict on the requirements.txt that required a upgrade of a min version of a dependency.
- I had to run `go mod tidy` in some of golang examples that made some changes.
- There is still a case of images and labels with `ghcr.io` that when done manually it adds the image to the right repository but keeps it private. This step of making images public and maybe attach them to this repo might not work in all case scenarios.

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -20,6 +20,9 @@ jobs:
         go-version: ["1.19"]
         platform: ["ubuntu-latest"]
     runs-on: ${{ matrix.platform }}
+    env:
+      KO_DOCKER_REPO: ghcr.io/vmware-samples/vcenter-event-broker-appliance # .../router@sha256:
+      REGISTRY: ghcr.io
```
Comment (embano1) [inline .github/workflows/release.yml:25]: nit: IMHO not needed

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -45,7 +48,7 @@ jobs:
           docker run --rm -v $PWD:/workdir ${IMAGE}@sha256:${IMAGE_SHA} -o vmware-event-router/RELEASE_CHANGELOG.md $(basename "${{ github.ref }}" )
 
       - name: GoReleaser
-        uses: goreleaser/goreleaser-action@5fdedb94abba051217030cc86d4523cf3f02243d
+        uses: goreleaser/goreleaser-action@7ec5c2b0c6cdda6e8bbb49444bc797dd33d74dd8
```
Comment (embano1) [inline .github/workflows/release.yml:51]: I suggest keeping the `4.6.0` commit for now unless you have tested the `5.0.0` change?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -100,11 +103,11 @@ jobs:
...
   images:
     needs: release
-    name: Create GCR image and release.yaml
+    name: Create Github Package image and release.yaml
     timeout-minutes: 30
     env:
-      KO_DOCKER_REPO: us.gcr.io/daisy-284300/veba # .../router@sha256:
-      PROJECT_ID: ${{ secrets.GCP_PROJECT_ID }}
+      KO_DOCKER_REPO: ghcr.io/vmware-samples/vcenter-event-broker-appliance # .../router@sha256:
+      REGISTRY: ghcr.io
```
Comment (embano1) [inline .github/workflows/release.yml:110]: nit: remove both since it's the default in `ko`

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -124,16 +127,14 @@ jobs:
           cache-dependency-path: vmware-event-router/go.sum
         id: go
         
-      - id: login_to_gcp
-        uses: google-github-actions/auth@35b0e87d162680511bf346c299f71c9c5c379033
+      - name: Log in to the Container registry
```
Comment (embano1) [inline .github/workflows/release.yml:130]: if you remove the `ENV` in this job, `ko` will default to `ghcr` and automatically log in https://github.com/ko-build/setup-ko/blob/main/action.yml

Comment (embano1) [reply .github/workflows/release.yml:130]: can you please change `imjasonh/setup-ko@ace48d793556083a76f1e3e6068850c1f4a369aa` to `ko-build/setup-ko@v0.6`?

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -124,16 +127,14 @@ jobs:
...
+        uses: docker/login-action@v1
```
Comment (embano1) [inline .github/workflows/release.yml:131]: can be removed, see above

File: .github/workflows/router-build.yml [hunk 6]
```diff
@@ -67,7 +67,7 @@ jobs:
     runs-on: ${{ matrix.platform }}
     timeout-minutes: 20
     env:
-      KO_DOCKER_REPO: us.gcr.io/daisy-284300/veba # .../router@sha256:
+      KO_DOCKER_REPO: ghcr.io/vmware-samples/vcenter-event-broker-appliance # .../router@sha256:
```
Comment (embano1) [inline .github/workflows/router-build.yml:70]: see comments in `release` workflow

File: .github/workflows/upload-dev-image.yml [hunk 7]
```diff
@@ -25,8 +25,8 @@ jobs:
     runs-on: ${{ matrix.platform }}
     timeout-minutes: 20
     env:
-      KO_DOCKER_REPO: us.gcr.io/daisy-284300/veba # .../router@sha256:
-      PROJECT_ID: ${{ secrets.GCP_PROJECT_ID }}
+      KO_DOCKER_REPO: ghcr.io/vmware-samples/vcenter-event-broker-appliance # .../router@sha256:
```
Comment (embano1) [inline .github/workflows/upload-dev-image.yml:28]: see comments in `release` workflow

File: .github/workflows/upload-dev-image.yml [hunk 8]
```diff
@@ -40,24 +40,20 @@ jobs:
...
           echo "KO_COMMIT=$(echo -n $GITHUB_SHA | cut -c -8)" >> $GITHUB_ENV
           echo "KO_TAG=$(basename "${{ github.ref }}")" >> $GITHUB_ENV
-
       - name: Build and push multi-arch image
+        id: build
         run: |
           # build and push with development tag
-          ko resolve --platform=linux/arm64,linux/amd64 --tags ${KO_COMMIT},development -BRf deploy/event-router-k8s.yaml
\ No newline at end of file
+          ko resolve --platform=linux/arm64,linux/amd64 --tags ${KO_COMMIT},${KO_TAG} -BRf deploy/event-router-k8s.yaml
```
Comment (embano1) [inline .github/workflows/upload-dev-image.yml:59]: why the change from `development` to `tag`?

File: .github/workflows/upload-prerelease-image.yml [hunk 9]
```diff
@@ -20,8 +20,8 @@ jobs:
     runs-on: ${{ matrix.platform }}
     timeout-minutes: 20
     env:
-      KO_DOCKER_REPO: us.gcr.io/daisy-284300/veba # .../router@sha256:
-      PROJECT_ID: ${{ secrets.GCP_PROJECT_ID }}
+      KO_DOCKER_REPO: ghcr.io/vmware-samples/vcenter-event-broker-appliance # .../router@sha256:
```
Comment (embano1) [inline .github/workflows/upload-prerelease-image.yml:23]: see comments in `release` workflow

File: examples/CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+## Publish a new version of an example
+
+These examples are hosted in [`ghcr.io`](https://github.com/features/packages) and we will explain how to deploy the examples to this registry but you are free to use any registry of your like.
+
+To publish an image manually you will need a [personal github token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens) (Classic) and save it in a environment variable like `$GH_CLASSIC_TOKEN`.
+
+After you have your token in your environment you can login using
+
+```bash
+echo $GH_CLASSIC | ko login ghcr.io -u <your-username> --password-stdin
```
Comment (embano1) [inline examples/CONTRIBUTING.md:30]:
```suggestion
echo $GH_CLASSIC_TOKEN | ko login ghcr.io -u <your-username> --password-stdin
```

File: examples/CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+```
+
+You also need to set two environment variables for `ko`: `KO_TAG` and `KO_COMMIT`
+
+```bash
+export KO_COMMIT=$(git rev-parse --short=8 HEAD)
+export KO_TAG=<new-version>
+```
+
+Now you can build and push your new version of the example using `ko build`
```
Comment (embano1) [inline examples/CONTRIBUTING.md:52]: Since you use the `ghcr.io/vmware-samples/vcenter-event-broker-appliance` repo throughout this guide (besides in the Python section btw), normal users will see an error during push due to missing permissions

File: vmware-event-router/.goreleaser.yaml [hunk 12]
```diff
@@ -29,3 +29,12 @@ snapshot:
 
 checksum:
   name_template: "checksums.txt"
+
+release:
+  github:
+    owner: vmware-samples
+    name: vcenter-event-broker-appliance
+  draft: true
```
Comment (embano1) [inline vmware-event-router/.goreleaser.yaml:37]: IMHO can be set to `false`

**Review by embano1 — CHANGES_REQUESTED** [review]

File: .github/workflows/upload-dev-image.yml (hunk 8 above)
Comment (dolfolife) [reply .github/workflows/upload-dev-image.yml:59]: the TAG will be `development` but I will revert this change.

File: examples/CONTRIBUTING.md (hunk 11 above)
Comment (dolfolife) [reply examples/CONTRIBUTING.md:52]: I try to use a generic way to use the package repository that will be use but let me know if you would add or delete something

**Comment by dolfolife** [comment]
@embano1 made some changes. I hope I understood your change requests, but if I miss something please send me a message. Happy to continue reviewing this.

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -32,7 +29,7 @@ jobs:
           ref: "master"
 
       - name: Set up Go
-        uses: actions/setup-go@93397bea11091df50f3d7e59dc26a7711a8bcfbe
+        uses: actions/setup-go@v0.6
```
Comment (embano1) [inline .github/workflows/release.yml:32]: change `setup-ko` not `setup-go` :D
https://github.com/ko-build/setup-ko

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -124,16 +121,10 @@ jobs:
...
-        uses: google-github-actions/auth@35b0e87d162680511bf346c299f71c9c5c379033
-        with:
-          credentials_json: '${{ secrets.GCP_CREDENTIALS }}'
-
-      - name: Set up gcloud
-        uses: google-github-actions/setup-gcloud@e30db14379863a8c79331b04a9969f4c1e225e0b
-
-      - name: Configure Docker
-        run: gcloud auth configure-docker --quiet
+      - name: Log in to the Container registry
```
Comment (embano1) [inline .github/workflows/release.yml:124]: remove

File: .github/workflows/release.yml [hunk 15]
```diff
@@ -124,16 +121,10 @@ jobs:
...
+        uses: docker/login-action@v1
+        env:
+          VES_DOCKER: /usr/bin/docker
```
Comment (embano1) [inline .github/workflows/release.yml:127]: remove

File: .github/workflows/router-build.yml [hunk 16]
```diff
@@ -40,7 +40,7 @@ jobs:
         id: go
         
       - name: GoReleaser Snapshot
-        uses: goreleaser/goreleaser-action@5fdedb94abba051217030cc86d4523cf3f02243d
+        uses: goreleaser/goreleaser-action@7ec5c2b0c6cdda6e8bbb49444bc797dd33d74dd8
```
Comment (embano1) [inline .github/workflows/router-build.yml:43]: inconsistent with `release` workflow

File: .github/workflows/upload-dev-image.yml [hunk 17]
```diff
@@ -40,24 +37,20 @@ jobs:
...
-      - name: Set up gcloud
-        uses: google-github-actions/setup-gcloud@e30db14379863a8c79331b04a9969f4c1e225e0b
-
-      - name: Configure Docker
-        run: gcloud auth configure-docker --quiet
-
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+        env:
```
Comment (embano1) [inline .github/workflows/upload-dev-image.yml:46]: remove everything

File: .github/workflows/upload-prerelease-image.yml [hunk 18]
```diff
@@ -35,24 +31,19 @@ jobs:
...
-        uses: google-github-actions/setup-gcloud@e30db14379863a8c79331b04a9969f4c1e225e0b
-
-      - name: Configure Docker
-        run: gcloud auth configure-docker --quiet
-
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+        env:
+          VES_DOCKER: /usr/bin/docker
```
Comment (embano1) [inline .github/workflows/upload-prerelease-image.yml:41]: remove everything

File: examples/CONTRIBUTING.md (hunk 11 above)
Comment (embano1) [reply examples/CONTRIBUTING.md:52]: You could just use `export KO_DOCKER_REPO=<your_docker_repo>`

File: examples/CONTRIBUTING.md [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+```bash
+ko build . --image-label org.opencontainers.image.source="https://github.com/<github-account>/vcenter-event-broker-appliance/"  --tags <new-version> --tags latest -B
+```
+
+### Notes
+- On GitHub, in order to attach an image to a repository we need to set the label `org.opencontainers.image.source` of the repository the package will be associated with.
+- we use the flag `-B` in order to use the base path without MD5 hash after `KO_DOCKER_REPO`
+- > Currently, there is a behavior I am trying to debug where the tag of the image source does not work and it does not attach the package to the repository. If you find yourself in the situation you can attach the repository on the GitHub Package webpage
```
Comment (embano1) [inline examples/CONTRIBUTING.md:61]: Do you want to keep this?

File: examples/CONTRIBUTING.md [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+
+## Powercli Powershell
+
+For the Powercli examples we need to build the template that is on [knative/templates](./knative/templates/)
+
+These templates use Dockerfiles and can build them using `docker build`. After you built the images we need to tag the images
+
+```bash
+docker image tag <template-image> ghcr.io/<github-account>/vcenter-event-broker-appliance/<template-image>:latest
```
Comment (embano1) [inline examples/CONTRIBUTING.md:70]: same comment as above on docker repo name

File: examples/CONTRIBUTING.md [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+> Notes: the PowerCLI template depends on the PowerPS image. You need to build the base before the CLI version
+
+After you build the templates, you can now build any example on `knative/powershell` or `knative/powercli` using the same steps of the templates
+
+## Python
+
+[Buildpacks](https://buildpacks.io) are used to create the container image. We can build any example by using:
+
+```bash
+IMAGE=<docker-username>/<example-name>:<version>
```
Comment (embano1) [inline examples/CONTRIBUTING.md:89]: see you're using a good example here :)

**Review by embano1 — CHANGES_REQUESTED** [review]

**Comment by embano1** [comment]
> @embano1 made some changes. I hope I understood your change requests, but if I miss something please send me a message. Happy to continue reviewing this.

Yeah, mostly the Docker login stuff which can go away in all workflows since we're using `ko` defaults

File: .github/workflows/release.yml (hunk 13 above)
Comment (dolfolife) [reply .github/workflows/release.yml:32]: oops

Comment (dolfolife) [reply .github/workflows/release.yml:32]: from the example, it still needs setup-go no?

File: .github/workflows/release.yml [hunk 22]
```diff
@@ -28,12 +28,15 @@ jobs:
...
 
-      - name: Set up Go
+      - name: Set up Go ${{ matrix.go-version }}
         uses: actions/setup-go@93397bea11091df50f3d7e59dc26a7711a8bcfbe
         with:
           go-version: ${{ matrix.go-version }}
           cache-dependency-path: vmware-event-router/go.sum
 
+      - name: Setup ko
+        uses: actions/setup-ko@v0.6
```
Comment (embano1) [inline .github/workflows/release.yml:38]: use correct repo and SHA (IIRC required by OSPO security)

`ko-build/setup-ko@@ace48d793556083a76f1e3e6068850c1f4a369aa`

File: .github/workflows/release.yml [hunk 23]
```diff
@@ -114,27 +114,16 @@ jobs:
...
-          credentials_json: '${{ secrets.GCP_CREDENTIALS }}'
-
-      - name: Set up gcloud
-        uses: google-github-actions/setup-gcloud@e30db14379863a8c79331b04a9969f4c1e225e0b
-
-      - name: Configure Docker
-        run: gcloud auth configure-docker --quiet
 
+      - name: Setup ko
+        uses: actions/setup-ko@v0.6
```
Comment (embano1) [inline .github/workflows/release.yml:125]: same as above

File: .github/workflows/router-build.yml [hunk 24]
```diff
@@ -66,15 +66,13 @@ jobs:
...
-    env:
-      KO_DOCKER_REPO: us.gcr.io/daisy-284300/veba # .../router@sha256:
 
     steps:
       - name: Check out code
         uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac
         
       - name: Setup ko
-        uses: imjasonh/setup-ko@ace48d793556083a76f1e3e6068850c1f4a369aa # v0.6
+        uses: actions/setup-ko@v0.6
```
Comment (embano1) [inline .github/workflows/router-build.yml:75]: use SHA (apply to all)

**Review by embano1 — CHANGES_REQUESTED** [review]

**Comment by embano1** [comment]
@dolfolife good job! Don't worry about the integration test (OpenFaaS): I'll fix later. Can you please squash your commits?

**Comment by embano1** [comment]
Is @lamw going to rebuild all images?

**Review by embano1 — APPROVED** [review]
Thx a ton @dolfolife
